src/collectives/all_to_all_api.cc

```diff
@@ -7,20 +7,28 @@
 
 #include "enqueue.h"
 #include "collectives.h"
+#include "graph/topo.h"
 
 NCCL_API(ncclResult_t, ncclAllToAll, const void* sendbuff, void* recvbuff, size_t count, ncclDataType_t datatype,
   ncclComm_t comm, hipStream_t stream);
 ncclResult_t ncclAllToAll(const void* sendbuff, void* recvbuff, size_t count, ncclDataType_t datatype,
   ncclComm_t comm, hipStream_t stream) {
-  int nRanks;
-  NCCLCHECK(ncclCommCount(comm, &nRanks));
-  size_t rankOffset = count * ncclTypeSize(datatype);
-  if (count == 0) return ncclSuccess;
-  NCCLCHECK(ncclGroupStart());
-  for (int r=0; r<nRanks; r++) {
-    NCCLCHECK(ncclSend(((char*)sendbuff)+r*rankOffset, count, datatype, r, comm, stream));
-    NCCLCHECK(ncclRecv(((char*)recvbuff)+r*rankOffset, count, datatype, r, comm, stream));
+  if (comm->topo->pivotA2AEnabled) {
```

Comment:
Decision to take pivot alltoall path needs to be "voted" using https://github.com/ROCmSoftwarePlatform/rccl/blob/develop/src/init.cc#L1044. This is to make sure all participating ranks are in sync. You can add a field in allGather3Data.

Reply:
Fixed.